### PR TITLE
ModalView: code cleanup regarding detection of main-Window:

### PR DESCRIPTION
**Pull request by AndiEcker**
#7412 -discussion-follow-up from 14-Mar-2021 with @matham (pinging @tshirtman).

- removed unneeded _search_window() method and attach_to property.
- added undocumented events on_pre_open and on_pre_dismiss to module docstring.
- allow stacking of multiple ModalView instances.
- simplified demo code.
- simplified unit test (test_uix_modal.py).

This is a WIP commit (1 of 2) - if you agree with the changes done in this first step, I'd like to go even further and replace the current `_window` property of ModalView with the main window `kivy.core.window.Window`, because `_window` is currently only used as a flag if the modal view is opened/visible (holding `None` if not opened or the main window instance if opened).

The only place where the window instance is accessed is in the declaration of ModalView (in style.kv), which could also be replaced with `Window`.

The opening state of the modal view could then be represented by an added `is_open` property (will be added in commit 2 of 2).

These changes would make the modal view faster, more readable/understandable, and easier to maintain/test.

Maintainer merge checklist
* [ ] Title is descriptive/clear for inclusion in release notes.
* [ ] Applied a `Component: xxx` label.
* [ ] Applied the `api-deprecation` or `api-break` label.
* [ ] Applied the `release-highlight` label to be highlighted in release notes.
* [ ] Added to the milestone version it was merged into.
* [ ] **Unittests** are included in PR.
* [ ] Properly documented, including `versionadded`, `versionchanged` as needed.

### CODE REVIEW

**Comment by AndiEcker** [comment]
@matham @tshirtman 

could you please review this PR.

The last/3rd commit did revert the changes in style.kv.
Had to restore the `_window` property because neither
`Window` nor `app` can be used in style.kv.

modalview.py is now 20 lines shorter and the
unit tests got 20 lines more.

File: kivy/uix/modalview.py
Comment (tshirtman) [inline kivy/uix/modalview.py:115]: it might not be an api break, because if anyone sets a value for this, the parent window will be the main one anyway, but do we need to document that the property now does nothing? It's unlikely that people bound anything to this property, but it would break their code if they did, so maybe leaving it in with a statement that it has no behavior now and will be removed in the future would be better?

**Review by tshirtman — APPROVED** [review]
Looks good to me, just not completely comfortable about the small api break, but it might be just fine.

File: kivy/uix/modalview.py
Comment (AndiEcker) [reply kivy/uix/modalview.py:115]: Thank you very much @tshirtman  for your review/approval!!! :+1: 

Getting your concerns and therefore just added (in the 4rd/last commit) a `versionchanged` to
the docstring of the modalview module regarding the removal of `attach_to` (instead of marking
it as deprecated, to keep the Kivy code base more compact and clean).

As you said setting/using this property will make no difference, so an old project using it could
easily be fixed by just removing any reference to `ModalView.attach_to` - hope you agree!!

File: kivy/uix/modalview.py
Comment (matham) [inline kivy/uix/modalview.py:76]:
```suggestion
.. versionchanged:: 2.1.0

    Removed `attach_to` property. In contrary to :attr:`DropDown.attach_to`
```

I think you need a line break there.

File: kivy/uix/modalview.py
Comment (matham) [inline kivy/uix/modalview.py:117]: While it does default to True, it should be explicitly provided. It just makes it harder to follow if the default isn't provided and doesn't really help. So please add it back.

File: kivy/uix/modalview.py
Comment (matham) [inline kivy/uix/modalview.py:176]: Please put back the default property values. There's no reason to remove them, so removing them can only contribute to bugs and it helps with code clarity being able to see the default value, even if we know what it defaults to when not provided. Same for all the other places.

File: kivy/uix/modalview.py
Comment (matham) [inline kivy/uix/modalview.py:282]: While technically your change is ok, I would feel better if you leave this unchanged. Because although `self._window` is `Window`, it is possible for the Window to have changed, e.g. during testing when we may create a new window for each test (although currently we actually re-use it). 

Also, on principle, it's better to remove/unbind from from `self._window` because we know that is the original object we use for sure, while Window, although very likely to also be the same object, ould possibly be changed in the future so it's less robust.

File: kivy/uix/modalview.py
Comment (matham) [inline kivy/uix/modalview.py:193]:
```suggestion
        """Display the modal in the Window.
```

File: kivy/uix/modalview.py
Comment (matham) [inline kivy/uix/modalview.py:201]:
```suggestion
        """
        if self._is_open:
            return
        self._window = Window
```
Don't we still need to check if it's already open?

File: kivy/uix/modalview.py
Comment (matham) [reply kivy/uix/modalview.py:115]: Yeah, I am similarly concerned by the API break.

And like @tshirtman suggested, it's better to leave the property in place saying it's not used anymore and remove it in a latter version. rather than abruptly dropping it. 

It can be done by adding `deprecated` arg as follows: `attach_to = ObjectProperty(None, deprecated=True)`. And update the docs that it is not yet dropped, but will be in the future.

**Review by matham — CHANGES_REQUESTED** [review]

File: kivy/uix/modalview.py
Comment (AndiEcker) [reply kivy/uix/modalview.py:76]: Thank you very much @matham for your review!!! Sphinx does not need a line break here (see e.g. [https://kivy.org/doc/stable/api-kivy.core.text.html](https://kivy.org/doc/stable/api-kivy.core.text.html).

File: kivy/uix/modalview.py
Comment (AndiEcker) [reply kivy/uix/modalview.py:117]: Ok, although I personally prefer short/clean code, especially in cases like this, where the default value is **explicitly** documented.

File: kivy/uix/modalview.py
Comment (AndiEcker) [reply kivy/uix/modalview.py:176]: Ok

File: kivy/uix/modalview.py
Comment (AndiEcker) [reply kivy/uix/modalview.py:201]: good catch

File: kivy/uix/modalview.py
Comment (AndiEcker) [reply kivy/uix/modalview.py:76]: Reworded and moved it to the `ModalView` class comment, also to resolve your other change request (keep the property, and mark it as deprecated).

**Review by tshirtman — APPROVED** [review]
LGTM

File: kivy/uix/modalview.py
Comment (matham) [inline kivy/uix/modalview.py:122]:
```suggestion
```

Can you please remove these? They shouldn't be in the repo. It's not required to pass our tests and this is a pycharm specific false-positive warning because this code is perfectly good python.

If we start accepting IDE specific flags, or even general flags soon they'll be all over the codebase. The point of linting is to help, not to become an additional piece of code that needs to be managed. I understand it's annoying to see these pycharm highlighting warnings, but they are meant to help and if the goal then becomes to make pycharm show no warnings then it kinda lost the point of being there to help write better code, not wrangle with the IDE.

I'll link to a video that @tshirtman loves to quote about this topic: https://www.youtube.com/watch?v=wf-BqAjZb8M.

File: kivy/uix/modalview.py
Comment (matham) [inline kivy/uix/modalview.py:183]:
```suggestion
```

File: kivy/uix/modalview.py
Comment (matham) [inline kivy/uix/modalview.py:258]:
```suggestion
            self.center = self._window.center
```

File: kivy/uix/modalview.py
Comment (matham) [inline kivy/uix/modalview.py:80]:
```suggestion
```
I glossed over this before, but generally we try not to import `Window` at the module level because that causes a Window to be created upon import. Sometimes it's unavoidable and I know that importing kivy will already generally create the window, but it's better to reduce this for now until we can figure out how to resolve this issue of Kivy creating a Window when Window is imported.

File: kivy/uix/modalview.py
Comment (matham) [inline kivy/uix/modalview.py:209]:
```suggestion
        """
        from kivy.core.window import Window
        if self._is_open:
```

File: kivy/uix/modalview.py
Comment (matham) [inline kivy/uix/modalview.py:318]:
```suggestion
    from kivy.uix.button import Button
    from kivy.core.window import Window
```

**Review by matham — CHANGES_REQUESTED** [review]

**Review by matham — APPROVED** [review]

**Comment by matham** [comment]
I hope you don't mind me merging the suggested changes, but I didn't want to make you wait for another review round. Thanks for your effort!

**Pull request merged**
